[Feat/notification readview] notification read view 작성

han-son-03 opened the pull request:
## ✅ PR 요약
- 작업 요약: notification read view 작성
- notification read test code

## 📄 상세 내용
- [ ]  notification read 관련 view 작성
- [ ] notification read 관련 test code 작성
- [ ] url 매핑
- [ ] 읽음 처리 테스트 에러 수정


## 📸 스크린샷 (선택)

## 📝 기타 참고 사항
- [ ] mypy cache 오류 해결

## 🧪 PR Checklist
- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

han-son-03 committed:
✨ feat     : notification read 관련 view 작성

사용자가 알림에 대한 단건과 해당하는 전체 알림에 대한 읽음 처리 api 입니다.

han-son-03 committed:
✅ test     : notification read 관련 test code 작성

읽음 처리 테스트와, 중복 읽음 처리, 없은 알림 에러처리, 전체 읽음 처리에 관한 테스트 코드 입니다

han-son-03 committed:
💡 chore    : url 경로 지정, api_urls 추가

test코드 작성 후, reverse viewname과 함께 url 작성

han-son-03 committed:
🐛 fix      : Notification 읽음 처리 테스트 에러 수정

- 로그인 처리 추가하여 AnonymousUser 문제 해결
- 테스트 함수에 -> None 반환 타입 추가하여 mypy 타입 체크 통과
- black, isort 적용

han-son-03 committed:
🐛 fix      : rebase 이후 urls.py 구조 통일

urls.py의 파일 위치를 rebase하며 원격 브랜치와 통일했습니다(기존: apps/notification/urls/api_urls.py 레퍼런스 참고 구조-> apps/notification/urls.py)

최종 Push 직전 mypy cache 문제로 다른 파트에서 오류가 있었습니다.

최종 Push 직전 다른 파트에서 mypy에 의한 cache 문제가 있었습니다

Meoyoug commented on apps/notification/tests/test_read_views.py:
```diff
@@ -0,0 +1,75 @@
+
+
+class NotificationReadAPITestCase(APITestCase):
+    def setUp(self) -> None:
```
테스트간 지속적으로 사용되는 자원이고 해당 자원에 대한 수정, 삭제가 일어나지 않고 읽어오기만 하는 자원이라면 setUpTestData를 활용하시는게 좋습니다.

setUp의 경우 이 테스트 케이스 클래스 내의 모든 테스트 함수가 실행되기전 한번씩 실행되지만 setUpTestData의 경우 이 테스트 클래스가 인스턴스화 되기전에 1번만 실행되서 불필요한 쿼리를 아낄 수 있어요

han-son-03 replied:
감사합니다

SWWS97 reviewed (changes requested):
현재 해당 유저의 알림이 아닐 경우에 대한 예외 처리가 없는거 같아요

SWWS97 commented on apps/notification/views/read_views.py:
```diff
@@ -0,0 +1,61 @@
+    def post(self, request: Request, notification_id: int) -> Response:
+        notification = self.get_object(notification_id)
+        if not notification:
+            return Response({"error": "해당 알림 내역을 찾을 수 없습니다."}, status=status.HTTP_404_NOT_FOUND)
```
404 에러에 대한 응답이 명세서와 다르게 되어있는거 같아요

2joonkim commented on apps/notification/tests/test_read_views.py:
```diff
@@ -0,0 +1,75 @@
+            user=self.user,
+            content="읽지 않은 알림",
+            is_read=False,
+        )
+        self.notification_read = Notification.objects.create(
+            user=self.user,
+            content="읽음 처리된 알림",
+            is_read=True,
+        )
```
현재 모델 보시면 옵션 하나 빠져있어요
back_url_link 이게 빠져있는거 같은데
현재 테스트코드 단에서는 아마 back_url_link="" 빈값으로 적용되어 통과는 될 겁니다
근데 스펙 API 테스트 목적이면 상관없지만 나중에 실제 서비스 테스트 작성할 때 필드 채우면 될 거 같습니다.

han-son-03 replied:
감사합니다

2joonkim commented on apps/notification/views/read_views.py:
```diff
@@ -0,0 +1,61 @@
+        tags=["notification"],
+        summary="로그인 유저가 수신한 알림중 읽지 않은 모든 알림을 읽음 처리하는 API 입니다.",
+        responses={
+            200: {"type": "object", "example": {"detail": "모든 알림 읽음처리에 성공하였습니다."}},
```
요거 스키마랑 실제 응답이랑 조금 다른데요?? 이거 통일시켜주셔야 할 거 같아요

han-son-03 replied:
확인해보겠습니다 감사합니다

han-son-03 committed:
💡 chore: 수정 사항 반영

test code 옵션 추가(back_url_link), 스키마 통일, test 코드 setUp -> setUpTestData로 반영하여 불필요한 쿼리 감소

han-son-03 committed:
🎨 style: black, isort 코드 포매터 실행

scripts/test.sh 실행하면서 괜찮을 줄 알았는데 ci isort에서 걸렸습니다 추가적으로 isort, black 재실행 후 push 합니다
